graphbolt/src/csc_sampling_graph.cc

```diff
@@ -636,11 +636,108 @@ inline int64_t NonUniformPick(
+          auto positive_probs_indices_ptr =
+              positive_probs_indices.data_ptr<PickedType>();
+
+          std::random_device rd;
+          std::mt19937 gen(rd());
+          if (!replace) {
+            // The algorithm is from gumbel softmax.
+            // s = argmax( logp - log(-log(eps)) ) where eps ~ U(0, 1)
+            // Here we can apply exp to the formula which will not affect result
+            // of argmax or topk. Then we have s = argmax( p / (-log(eps)) )
+            // where eps ~ U(0, 1). We can also simplify the formula above by s
+            // = argmax( p / q ) where q ~ Exp(1)
+            std::exponential_distribution<> exp_distrib(1);
+            if (fanout == 1) {
+              // Return argmax(p / q).
+              scalar_t max_prob = 0;
+              PickedType max_prob_index = -1;
+              for (auto i = 0; i < num_positive_probs; ++i) {
+                scalar_t cur_prob =
+                    local_probs_data_ptr[positive_probs_indices_ptr[i]] /
+                    exp_distrib(gen);
+                if (cur_prob > max_prob) {
+                  max_prob = cur_prob;
+                  max_prob_index = positive_probs_indices_ptr[i];
+                }
+              }
+              *picked_data_ptr = max_prob_index + offset;
+            } else {
+              // Return topk(p / q).
+              std::vector<std::pair<scalar_t, PickedType>> q(
+                  num_positive_probs);
+              for (auto i = 0; i < num_positive_probs; ++i) {
+                q[i].first =
+                    local_probs_data_ptr[positive_probs_indices_ptr[i]] /
+                    exp_distrib(gen);
+                q[i].second = positive_probs_indices_ptr[i];
+              }
+              if (fanout < num_positive_probs / 64) {
+                // Use partial_sort.
+                std::partial_sort(
+                    q.begin(), q.begin() + fanout, q.end(), std::greater{});
+                for (auto i = 0; i < fanout; ++i) {
+                  picked_data_ptr[i] = q[i].second + offset;
+                }
+              } else {
+                // Use nth_element.
+                std::nth_element(
+                    q.begin(), q.begin() + fanout - 1, q.end(), std::greater{});
+                for (auto i = 0; i < fanout; ++i) {
+                  picked_data_ptr[i] = q[i].second + offset;
+                }
+              }
+            }
+          } else {
+            // Calculate cumulative sum of probabilities.
+            std::vector<scalar_t> cum_probs(num_positive_probs);
+            scalar_t sum_probs = 0;
+            for (auto i = 0; i < num_positive_probs; ++i) {
+              sum_probs += local_probs_data_ptr[positive_probs_indices_ptr[i]];
+              cum_probs[i] = sum_probs;
+            }
+            // Normalize.
+            if ((sum_probs > 0) ||
+                ((sum_probs < 1.00001) && (sum_probs > 0.99999))) {
```

Comment:
The set of (sum_probs < 1.00001) && (sum_probs > 0.99999) is already included in the set of (sum_probs > 0), why we need ||?

Reply:
I actually copied and pasted this from ATen. I'm now a little confused. I don't think it is what their developers intended to write LOL.